src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,182 @@
+                                                                                      s_localizableDescription,
+                                                                                      isPortedFxCopRule: false,
+                                                                                      isDataflowRule: false);
+
+        private const string RequiresPreviewFeaturesAttribute = nameof(RequiresPreviewFeaturesAttribute);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        private bool thisAssemblyUsesPreviewFeatures;
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(context =>
+            {
+                IAssemblySymbol? thisAssembly = context.Compilation.Assembly;
+                ImmutableArray<AttributeData> attributes = thisAssembly.GetAttributes();
+
+                AttributeData? attribute = attributes.FirstOrDefault(x => x.AttributeClass.ToString() == RequiresPreviewFeaturesAttribute);
+                if (!(attribute is null))
+                {
+                    // This assembly has enabled preview attributes.
+                    return;
+                }
+
+                var enabledPreviewFeatures = context.Options.GetMSBuildPropertyValue(MSBuildPropertyOptionNames.EnablePreviewFeatures, context.Compilation, context.CancellationToken);
+                if (string.Equals(enabledPreviewFeatures, "true", System.StringComparison.OrdinalIgnoreCase))
+                {
+                    return;
+                }
+
+                ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols = new ConcurrentDictionary<ISymbol, IOperation?>();
+                context.RegisterOperationBlockStartAction(context => BuildSymbolInformation(context, requiresPreviewFeaturesSymbols));
+            });
+        }
+
+        private void BuildSymbolInformation(OperationBlockStartAnalysisContext context, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols)
+        {
+            context.RegisterOperationAction(context => AnalyzeOperation(context.Operation, requiresPreviewFeaturesSymbols),
+                OperationKind.Invocation,
+                OperationKind.ObjectCreation,
+                OperationKind.PropertyReference,
+                OperationKind.FieldReference,
+                OperationKind.DelegateCreation,
+                OperationKind.EventReference
+                );
+
+            context.RegisterOperationBlockEndAction(context =>
+            {
+                if (thisAssemblyUsesPreviewFeatures)
+                {
+                    foreach (var symbolAndOperation in requiresPreviewFeaturesSymbols)
```

Comment:
This is a problematic approach. This dictionary is created in the compilation start action and passed to every operation block start action so it will have new entries whenever a new method is analyzed. The dictionary's key-value pairs are then iterated in each operation block end, which is problematic as an entry added to this dictionary for the first operation block will be reported for every operation block end callback for every method in the compilation. Iterating a compilation wide mutable dictionary in an operation end block callback is not the right approach.

I think you should sync offline with @buyaa-n on the approach she used in platform compat analyzer for similar problem, similar approach might work for you.

Reply:
I have discussed this analyzer before with @pgovind, that time I thought it will be very similar to PCA but from this implementation and reading the design doc it seems not quite similar to PCA, at least it doesn't look like has any guard operation, therefore, no need flow analysis. If it doesn't need flow analysis i don't think it needs to register an action for `OperationBlockStartAction` and `rOperationBlockEndAction`, seems it only need to analyze `OperationAction`.

Reply:
Looks like you can report diagnostics immediately whenever it is found in the `OperationAction` as no further analysis  needed for the findings while using the cache only for searching an operation or adding a new one